service/history/workflowExecutionUtil.go

```diff
@@ -25,6 +25,74 @@ import (
 	"github.com/uber/cadence/common"
 )
 
+type workflowContext interface {
+	getContext() workflowExecutionContext
+	getMutableState() mutableState
+	reloadMutableState() (mutableState, error)
+	getReleaseFn() releaseWorkflowExecutionFunc
+	getRunID() string
+}
+
+type workflowContextImpl struct {
+	context      workflowExecutionContext
+	mutableState mutableState
+	releaseFn    releaseWorkflowExecutionFunc
+}
+
+type updateWorkflowAction struct {
+	noop           bool
+	createDecision bool
+}
+
+var (
+	emptyUpdateAction             = updateWorkflowAction{}
+	updateWorkflowWithNewDecision = updateWorkflowAction{
+		createDecision: true,
+	}
+	updateWorkflowWithoutDecision = updateWorkflowAction{
+		createDecision: false,
+	}
+)
+
+type updateWorkflowActionFunc func(mutableState mutableState) (updateWorkflowAction, error)
+
+func (w *workflowContextImpl) getContext() workflowExecutionContext {
+	return w.context
+}
+
+func (w *workflowContextImpl) getMutableState() mutableState {
+	return w.mutableState
+}
+
+func (w *workflowContextImpl) reloadMutableState() (mutableState, error) {
+	mutableState, err := w.getContext().loadWorkflowExecution()
+	if err != nil {
+		return nil, err
+	}
+	w.mutableState = mutableState
+	return mutableState, nil
+}
+
+func (w *workflowContextImpl) getReleaseFn() releaseWorkflowExecutionFunc {
+	return w.releaseFn
+}
+
+func (w *workflowContextImpl) getRunID() string {
+	return w.getContext().getExecution().GetRunId()
+}
+
+func newWorkflowContext(
+	context workflowExecutionContext,
+	releaseFn releaseWorkflowExecutionFunc,
+	mutableState mutableState) workflowContext {
```

Comment:
```go
func newWorkflow(
	context workflowExecutionContext,
	releaseFn releaseWorkflowExecutionFunc,
	mutableState mutableState,
)*workflowImpl{
```

Reply:
Why impl instead of interface? I think `newXXX` return interface is easier for future change (say change impl in future)

Reply:
no this is just the return value, easier for writing unit test

Reply:
alright, updated